rpm/0038-sailfishos-embedlite-egl-Fix-mesa-egl-display-and-bu.patch

```diff
@@ -0,0 +1,488 @@
++  MOZ_ASSERT(interface);
++  return (EGLDisplay)(interface->nativeResourceForIntegration(QByteArrayLiteral("egldisplay")));
++#else
++  return EGL_NO_DISPLAY;
++#endif
++}
++
++#ifdef MOZ_WIDGET_QT
++#define WL_COMPOSITOR_CREATE_SURFACE 0
++#define WL_SURFACE_DESTROY 0
++
++// See /usr/include/wayland-egl-core.h
++static struct wl_egl_window * (*_wl_egl_window_create)(struct wl_surface *surface, int width, int height) = nullptr;
++static void (*_wl_egl_window_destroy)(struct wl_egl_window *egl_window) = nullptr;
++
++// See /usr/include/wayland-client-protocol.h
++static const struct wl_interface *_wl_surface_interface = nullptr;
++
++// See wayland/src/wayland-client-core.h
++static struct wl_proxy * (*_wl_proxy_marshal_constructor)(struct wl_proxy *proxy, uint32_t opcode, const struct wl_interface *interface, ...) = nullptr;
++static void (*_wl_proxy_marshal)(struct wl_proxy *p, uint32_t opcode, ...) = nullptr;
++static void (*_wl_proxy_destroy)(struct wl_proxy *proxy) = nullptr;
++
++static inline struct wl_surface * _wl_compositor_create_surface(struct wl_compositor *wl_compositor)
++{
++  MOZ_ASSERT(_wl_proxy_marshal_constructor);
++  struct wl_proxy *id;
++  id = _wl_proxy_marshal_constructor((struct wl_proxy *) wl_compositor, WL_COMPOSITOR_CREATE_SURFACE, _wl_surface_interface, NULL);
++  return (struct wl_surface *) id;
++}
++
++static inline void _wl_surface_destroy(struct wl_surface *wl_surface)
++{
++  MOZ_ASSERT(_wl_proxy_marshal);
++  MOZ_ASSERT(_wl_proxy_destroy);
++  _wl_proxy_marshal((struct wl_proxy *) wl_surface, WL_SURFACE_DESTROY);
++  _wl_proxy_destroy((struct wl_proxy *) wl_surface);
++}
++
++static bool LoadWaylandFunctions() {
++  if (_wl_proxy_marshal_constructor) {
++    // We load everything or nothing
++    return true;
++  }
++
++  void *handle;
++
++  // wl_egl_window_create
++  // wl_egl_window_destroy
++  handle = dlopen("libwayland-egl.so.1", RTLD_LAZY);
++  if (!handle) {
++    NS_WARNING("Error opening libwayland-egl.so.1");
++    return false;
++  }
++  *(void **)(&_wl_egl_window_create) = dlsym(handle, "wl_egl_window_create");
++  *(void **)(&_wl_egl_window_destroy) = dlsym(handle, "wl_egl_window_destroy");
++
++  dlclose(handle);
++
++  if (!_wl_egl_window_create || !_wl_egl_window_destroy) {
++    NS_WARNING("Error loading functions from libwayland-egl.so.1");
++    return false;
++  }
++
++  // wl_surface_interface
++  // wl_proxy_marshal_constructor
++  // wl_proxy_marshal, wl_proxy_destroy
++  handle = dlopen("libwayland-client.so.0", RTLD_LAZY);
++  if (!handle) {
++    NS_WARNING("Error opening libwayland-client.so.0");
++    return false;
++  }
++  _wl_surface_interface = (const struct wl_interface *)dlsym(handle, "wl_surface_interface");
++
++  *(void **)(&_wl_proxy_marshal_constructor) = dlsym(handle, "wl_proxy_marshal_constructor");
++  *(void **)(&_wl_proxy_marshal) = dlsym(handle, "wl_proxy_marshal");
++  *(void **)(&_wl_proxy_destroy) = dlsym(handle, "wl_proxy_destroy");
++
++  dlclose(handle);
```

Comment:
It's probably not too big of an issue, but if libwayland-egl.so.1 gets unloaded by a final dlclose (potentially somewhere else) and one of the wayland functions is then accessed this will crash. I think it might be better to move the dlclose after we are sure that the functions won't be accessed anymore. However i think it's probably unlikely that this will cause any issues so i'm fine with it either way.

Reply:
It's going to fail immediately because nothing else loads the library, or never because the wayland QPA plugin will link against it and that won't get unloaded ever.

Reply:
I pushed some changes to avoid the former happening (which may never have happened, but better safe than sorry?). The new changes are in a separate commit which I'll squash before merging.

Now, if the functions get dynamically loaded, they don't get unloaded again until all of the surfaces have been released, which in practice only happens during shutdown of the app.